ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -234,30 +172,35 @@ public HelixClusterManager(ClusterMapConfig clusterMapConfig, String instanceNam
       helixClusterManagerMetrics.initializePartitionMetrics();
       helixClusterManagerMetrics.initializeCapacityMetrics();
     }
-    localDatacenterId = dcToDcZkInfo.get(clusterMapConfig.clusterMapDatacenterName).dcZkInfo.getDcId();
+    localDatacenterId = dcToDcInfo.get(clusterMapConfig.clusterMapDatacenterName).dcZkInfo.getDcId();
     partitionSelectionHelper =
         new PartitionSelectionHelper(helixClusterManagerCallback, clusterMapConfig.clusterMapDatacenterName,
             clusterMapConfig.clustermapWritablePartitionMinReplicaCount);
     // register partition selection helper as a listener of cluster map changes.
-    dcToDcZkInfo.values()
-        .forEach(info -> info.clusterChangeHandler.registerClusterMapListener(partitionSelectionHelper));
+    registerClusterMapListener(partitionSelectionHelper);
   }
 
   /**
    * Initialize HelixManager in local datacenter and complete subscription of HelixPropertyStore to listen for
-   * PartitionOverride zNode.
+   * PartitionOverride zNode. This needs to happen before other datacenters are initialized so that any partition
+   * overrides can be properly honored.
    * @param dataCenterToZkAddress the map mapping each datacenter to its corresponding ZkAddress.
    * @param instanceName the String representation of the instance associated with this manager.
    * @param helixFactory the factory class to construct and get a reference to a {@link HelixManager}.
-   * @return the HelixManager of local datacenter
+   * @return the HelixManager of local datacenter, or {@code null} if the local datacenter is
+   *         {@link ReplicaType#CLOUD_BACKED}, as we currently do not support getting cluster state from Helix for cloud
+   *         datacenters.
    * @throws Exception
    */
   private HelixManager initializeHelixManagerAndPropertyStoreInLocalDC(Map<String, DcZkInfo> dataCenterToZkAddress,
       String instanceName, HelixFactory helixFactory) throws Exception {
     DcZkInfo dcZkInfo = dataCenterToZkAddress.get(clusterMapConfig.clusterMapDatacenterName);
     String zkConnectStr = dcZkInfo.getZkConnectStr();
-    HelixManager manager;
-    manager = helixFactory.getZKHelixManager(clusterName, instanceName, InstanceType.SPECTATOR, zkConnectStr);
+    if (dcZkInfo.getReplicaType() == ReplicaType.CLOUD_BACKED) {
+      return null;
+    }
```

Comment:
I am not quite sure returning null for cloud dc is reasonable here.  For now, it makes some sense but in the future we need to reconsider. 
Let's say a frontend is brought up in cloud dc and this method should populate `partitionOverrideInfoMap` for the frontend. With current logic, `partitionOverrideInfoMap` is empty and frontend may route PUT to those partitions which should be temporarily sealed.
If frontend in cloud will be initialized with different code path, then just ignore this comment.

Reply:
Where should a frontend in a datacenter without a helix ambry-server cluster get partitionOverrideInfo from? Some other datacenter? VCR cluster?

I think that not having partition override is not a blocker for to start testing so I would prefer to keep this as is. In the future, I feel that storing partition override info in the VCR cluster would be the right way forward, since configuring an alternate zk to get this from seems like a hack. However, that would require the extra dev effort to support spectating VCR clusters described in some of the other comments.

Reply:
Correct, VCR cluster is the right place to store partition override info.  (I agree this is not a blocker now, we can keep this in mind if we are doing cluster expansion in the future. Previously, I just want to mention a case where some partitions are administratively sealed in linkedin dc. However, if frontends in cloud still believe them unsealed, the PUT requests are still able to write into these partitions via cloud frontend)